File: apps/plugins/vscode/src/controlledFileRead.ts
import * as crypto from "node:crypto";
import * as fs from "node:fs/promises";
import * as path from "node:path";
import { TextDecoder } from "node:util";
import { bridgeVersion } from "./identity";

export type ControlledFileReadGuiMessage = {
  version: string;
  type: "gui.controlledAgentFileReadRequest";
  requestId?: string;
  payload?: Record<string, unknown>;
};

export type ControlledFileReadHostMessage = {
  version: string;
  type: "host.controlledAgentFileReadResult";
  requestId: string;
  payload: ControlledFileReadMetadata;
};

type ControlledFileReadRequest = {
  requestId: string;
  requestIdMintedBy: "gui" | "host";
  source: "gui" | "host";
  assistantMinted: false;
  controlledWorkspaceId: string;
  runId: string;
  workspaceRelativePath: string;
  maxBytes: number;
  maxLines: number;
  allowBody: boolean;
  singleFileOnly: true;
  recursive: false;
  globAllowed: false;
  regexAllowed: false;
  indexingAllowed: false;
};

type ControlledFileReadBlockedReason = "read_disabled" | "policy_denied" | "unsafe_path" | "outside_workspace" | "hidden_path" | "dependency_path" | "generated_path" | "binary_file" | "symlink_denied" | "too_large" | "budget_exceeded";

type ControlledFileReadMetadata = {
  kind: "controlled_agent_file_read";
  version: "2026-06-29";
  authority: "bounded_text_file_read";
  cloudRequired: false;
  executionAllowed: false;
  agentStartAllowed: false;
  workspace: {
    controlledWorkspaceId: string;
    runId: string;
    workspaceMode: "worktree";
    host: "vscode";
    privatePathExposed: false;
    workspaceLabel: "Controlled worktree";
  };
  request: {
    requestId: string;
    source: "gui" | "host";
    requestIdMintedBy: "gui" | "host";
    assistantMinted: false;
    workspaceRelativePath: string;
    textOnly: true;
    maxBytes: number;
    budget: {
      scope: "single_explicit_file";
      maxBytes: number;
      maxLines: number;
      allowBody: boolean;
      singleFileOnly: true;
      recursive: false;
      globAllowed: false;
      regexAllowed: false;
      indexingAllowed: false;
      budgetLabel: string;
    };
    requestedAt: string;
    reason: string;
  };
  policyFlags: {
    fileReadAllowed: boolean;
    fileWriteAllowed: false;
    shellAllowed: false;
    gitAllowed: false;
    providerAllowed: false;
    toolAllowed: false;
    hiddenSearchAllowed: false;
    indexingAllowed: false;
    binaryReadAllowed: false;
    symlinkAllowed: false;
    autoStartAllowed: false;
    autoApplyAllowed: false;
    autoRunAllowed: false;
  };
  result: {
    status: "blocked" | "success" | "truncated";
    cloudRequired: false;
    executionAllowed: false;
    bodyIncluded: boolean;
    truncated: boolean;
    sanitizedPathLabel?: string;
    byteCount?: number;
    lineCount?: number;
    contentHash?: string;
    text?: string;
    blockedReason?: ControlledFileReadBlockedReason;
    message: string;
  };
};

const maxControlledFileReadBytes = 8192;
const maxControlledFileReadLines = 240;
const safeIdPattern = /^[A-Za-z0-9][A-Za-z0-9._-]{0,79}$/;
const safeRequestIdPattern = /^[A-Za-z0-9][A-Za-z0-9_.-]{0,127}$/;
const dependencySegments = new Set(["node_modules", "vendor"]);
const generatedSegments = new Set(["dist", "build", "out", "target", "coverage", "generated", "tmp", "temp"]);
const secretNamePattern = /auth|credential|password|secret|token|access[_-]?token|api[_-]?key|^\.env$/i;
const secretTextPattern = /authorization|bearer|api[_-]?key|access[_-]?token|auth[_-]?token|secret|password|credential|cookie|BEGIN [A-Z ]*PRIVATE KEY|sk-(?:proj-)?[A-Za-z0-9_-]{8,}/i;

export function isControlledFileReadGuiMessage(value: unknown): value is ControlledFileReadGuiMessage {
  return parseControlledFileReadRequest(value) !== undefined;
}

export function isInvalidControlledFileReadRequestMessage(value: unknown): value is ControlledFileReadGuiMessage & { requestId: string } {
  if (!isPlainRecord(value)) {
    return false;
  }
  return hasOnlyKeys(value, ["version", "type", "requestId", "payload"]) &&
    value.version === bridgeVersion &&
    value.type === "gui.controlledAgentFileReadRequest" &&
    isRequiredRequestId(value.requestId) &&
    parseControlledFileReadRequest(value) === undefined;
}

export async function runControlledFileReadRequest(message: ControlledFileReadGuiMessage, workspaceRoots: readonly string[]): Promise<ControlledFileReadHostMessage> {
  const parsed = parseControlledFileReadRequest(message);
  if (!parsed) {
    const requestId = isRequiredRequestId(message.requestId) ? message.requestId : "invalid-request";
    return createControlledFileReadHostMessage(createFallbackRequest(requestId), "blocked", "policy_denied", "Host policy denied the explicit bounded read");
  }
  if (!parsed.allowBody) {
    return createControlledFileReadHostMessage(parsed, "blocked", "read_disabled", "Host policy denied the explicit bounded read");
  }
  const pathCheck = validateControlledWorkspaceRelativePath(parsed.workspaceRelativePath);
  if (pathCheck !== "ok") {
    return createControlledFileReadHostMessage(parsed, "blocked", pathCheck, "Host policy denied the explicit bounded read");
  }
  const resolved = await resolveControlledWorkspaceFile(parsed.workspaceRelativePath, workspaceRoots);
  if (!resolved.ok) {
    return createControlledFileReadHostMessage(parsed, "blocked", resolved.reason, "Host policy denied the explicit bounded read");
  }
  if (resolved.size > maxControlledFileReadBytes) {
    return createControlledFileReadHostMessage(parsed, "blocked", "too_large", "Host policy denied the explicit bounded read");
  }

  let bytes: Uint8Array;
  try {
    bytes = await fs.readFile(resolved.filePath);
  } catch {
    return createControlledFileReadHostMessage(parsed, "blocked", "policy_denied", "Host policy denied the explicit bounded read");
  }
  if (bytes.byteLength !== resolved.size || bytes.byteLength > maxControlledFileReadBytes || isBinaryBytes(bytes)) {
    return createControlledFileReadHostMessage(parsed, "blocked", bytes.byteLength > maxControlledFileReadBytes ? "too_large" : "binary_file", "Host policy denied the explicit bounded read");
  }

  let text: string;
  try {
    text = new TextDecoder("utf-8", { fatal: true }).decode(bytes);
  } catch {
    return createControlledFileReadHostMessage(parsed, "blocked", "binary_file", "Host policy denied the explicit bounded read");
  }
  if (hasBinaryLikeText(text) || secretTextPattern.test(text)) {
    return createControlledFileReadHostMessage(parsed, "blocked", hasBinaryLikeText(text) ? "binary_file" : "policy_denied", "Host policy denied the explicit bounded read");
  }

  const limited = limitText(text, parsed.maxBytes, parsed.maxLines);
  return createControlledFileReadHostMessage(parsed, limited.truncated ? "truncated" : "success", undefined, limited.truncated ? "Bounded text read returned a truncated preview" : "Bounded text read completed within budget", limited.text, limited.byteCount, limited.lineCount);
}

export function parseControlledFileReadRequest(value: unknown): ControlledFileReadRequest | undefined {
  if (!isPlainRecord(value) || !hasOnlyKeys(value, ["version", "type", "requestId", "payload"]) || value.version !== bridgeVersion || value.type !== "gui.controlledAgentFileReadRequest" || !isRequiredRequestId(value.requestId) || !isPlainRecord(value.payload)) {
    return undefined;
  }
  const payload = value.payload;
  if (!hasOnlyKeys(payload, ["requestIdMintedBy", "source", "assistantMinted", "controlledWorkspaceId", "runId", "runtimeSessionId", "sessionId", "workspaceRelativePath", "maxBytes", "maxLines", "allowBody", "singleFileOnly", "recursive", "globAllowed", "regexAllowed", "indexingAllowed"]) || payload.requestIdMintedBy !== "gui" || payload.source !== "gui" || payload.assistantMinted !== false || !isSafeId(payload.controlledWorkspaceId) || !isSafeId(payload.runId) || (payload.runtimeSessionId !== undefined && !isSafeId(payload.runtimeSessionId)) || (payload.sessionId !== undefined && !isSafeId(payload.sessionId)) || typeof payload.workspaceRelativePath !== "string" || !boundedInteger(payload.maxBytes, 1, maxControlledFileReadBytes) || !boundedInteger(payload.maxLines, 1, maxControlledFileReadLines) || typeof payload.allowBody !== "boolean" || payload.singleFileOnly !== true || payload.recursive !== false || payload.globAllowed !== false || payload.regexAllowed !== false || payload.indexingAllowed !== false) {
    return undefined;
  }
  if (validateControlledWorkspaceRelativePath(payload.workspaceRelativePath) !== "ok") {
    return undefined;
  }
  return {
    requestId: value.requestId,
    requestIdMintedBy: payload.requestIdMintedBy,
    source: payload.source,
    assistantMinted: payload.assistantMinted,
    controlledWorkspaceId: payload.controlledWorkspaceId,
    runId: payload.runId,
    workspaceRelativePath: payload.workspaceRelativePath,
    maxBytes: payload.maxBytes,
    maxLines: payload.maxLines,
    allowBody: payload.allowBody,
    singleFileOnly: payload.singleFileOnly,
    recursive: payload.recursive,
    globAllowed: payload.globAllowed,
    regexAllowed: payload.regexAllowed,
    indexingAllowed: payload.indexingAllowed,
  };
}

function createControlledFileReadHostMessage(request: ControlledFileReadRequest, status: "blocked" | "success" | "truncated", blockedReason: ControlledFileReadBlockedReason | undefined, message: string, text?: string, byteCount?: number, lineCount?: number): ControlledFileReadHostMessage {
  const bodyIncluded = text !== undefined && status !== "blocked";
  return {
    version: bridgeVersion,
    type: "host.controlledAgentFileReadResult",
    requestId: request.requestId,
    payload: {
      kind: "controlled_agent_file_read",
      version: "2026-06-29",
      authority: "bounded_text_file_read",
      cloudRequired: false,
      executionAllowed: false,
      agentStartAllowed: false,
      workspace: {
        controlledWorkspaceId: request.controlledWorkspaceId,
        runId: request.runId,
        workspaceMode: "worktree",
        host: "vscode",
        privatePathExposed: false,
        workspaceLabel: "Controlled worktree",
      },
      request: {
        requestId: request.requestId,
        source: request.source,
        requestIdMintedBy: request.requestIdMintedBy,
        assistantMinted: false,
        workspaceRelativePath: request.workspaceRelativePath,
        textOnly: true,
        maxBytes: request.maxBytes,
        budget: {
          scope: "single_explicit_file",
          maxBytes: request.maxBytes,
          maxLines: request.maxLines,
          allowBody: request.allowBody,
          singleFileOnly: true,
          recursive: false,
          globAllowed: false,
          regexAllowed: false,
          indexingAllowed: false,
          budgetLabel: "Small explicit text read budget",
        },
        requestedAt: new Date().toISOString().replace(/\.\d{3}Z$/, "Z"),
        reason: bodyIncluded ? "User approved one bounded text read for review" : "Host blocked this explicit read before returning text",
      },
      policyFlags: {
        fileReadAllowed: true,
        fileWriteAllowed: false,
        shellAllowed: false,
        gitAllowed: false,
        providerAllowed: false,
        toolAllowed: false,
        hiddenSearchAllowed: false,
        indexingAllowed: false,
        binaryReadAllowed: false,
        symlinkAllowed: false,
        autoStartAllowed: false,
        autoApplyAllowed: false,
        autoRunAllowed: false,
      },
      result: {
        status,
        cloudRequired: false,
        executionAllowed: false,
        bodyIncluded,
        truncated: status === "truncated",
        ...(bodyIncluded ? {
          sanitizedPathLabel: request.workspaceRelativePath,
          byteCount: byteCount ?? 0,
          lineCount: lineCount ?? 0,
          contentHash: `sha256:${crypto.createHash("sha256").update(text).digest("hex")}`,
          text,
        } : { blockedReason: blockedReason ?? "policy_denied" }),
        message,
      },
    },
  };
}

function createFallbackRequest(requestId: string): ControlledFileReadRequest {
  return {
    requestId,
    requestIdMintedBy: "gui",
    source: "gui",
    assistantMinted: false,
    controlledWorkspaceId: "workspace-read-blocked",
    runId: "run-read-blocked",
    workspaceRelativePath: "blocked/read.txt",
    maxBytes: 1,
    maxLines: 1,
    allowBody: false,
    singleFileOnly: true,
    recursive: false,
    globAllowed: false,
    regexAllowed: false,
    indexingAllowed: false,
  };
}

function validateControlledWorkspaceRelativePath(value: string): "ok" | ControlledFileReadBlockedReason {
  if (value.length === 0 || value.length > 180 || value.startsWith("/") || value.startsWith("~") || /^[A-Za-z]:/.test(value) || value.includes("\\") || value.includes(":") || value.includes("?") || value.includes("#") || value.includes("%") || /[\u0000-\u001f\u007f-\u009f]/.test(value) || value.includes("//") || value.endsWith("/")) {
    return "unsafe_path";
  }
  const segments = value.split("/");
  for (const segment of segments) {
    if (segment.length === 0 || segment === "." || segment === "..") {
      return "unsafe_path";
    }
    if (segment.startsWith(".")) {
      return "hidden_path";
    }
    if (dependencySegments.has(segment)) {
      return "dependency_path";
    }
    if (generatedSegments.has(segment)) {
      return "generated_path";
    }
    if (secretNamePattern.test(segment)) {
      return "unsafe_path";
    }
    if (!/^[A-Za-z0-9][A-Za-z0-9._-]*$/.test(segment)) {
      return "unsafe_path";
    }
  }
  return "ok";
}

async function resolveControlledWorkspaceFile(workspaceRelativePath: string, workspaceRoots: readonly string[]): Promise<{ ok: true; filePath: string; size: number } | { ok: false; reason: ControlledFileReadBlockedReason }> {
  const segments = workspaceRelativePath.split("/");
  const matches: { filePath: string; size: number }[] = [];
  for (const root of workspaceRoots) {
    const match = await resolveUnderRoot(root, segments);
    if (match.ok) {
      matches.push({ filePath: match.filePath, size: match.size });
    } else if (match.reason === "symlink_denied" || match.reason === "binary_file" || match.reason === "too_large") {
      return match;
    }
  }
  if (matches.length !== 1) {
    return { ok: false, reason: matches.length === 0 ? "outside_workspace" : "policy_denied" };
  }
  return { ok: true, filePath: matches[0].filePath, size: matches[0].size };
}

async function resolveUnderRoot(root: string, segments: string[]): Promise<{ ok: true; filePath: string; size: number } | { ok: false; reason: ControlledFileReadBlockedReason }> {
  let rootReal: string;
  try {
    rootReal = await fs.realpath(root);
  } catch {
    return { ok: false, reason: "outside_workspace" };
  }
  let current = root;
  for (const segment of segments) {
    current = path.join(current, segment);
    let entry;
    try {
      entry = await fs.lstat(current);
    } catch {
      return { ok: false, reason: "outside_workspace" };
    }
    if (entry.isSymbolicLink()) {
      return { ok: false, reason: "symlink_denied" };
    }
  }
  let stat;
  try {
    stat = await fs.stat(current);
  } catch {
    return { ok: false, reason: "outside_workspace" };
  }
  if (!stat.isFile()) {
    return { ok: false, reason: "outside_workspace" };
  }
  const fileReal = await fs.realpath(current);
  const relative = path.relative(rootReal, fileReal);
  if (relative.length === 0 || relative.startsWith("..") || path.isAbsolute(relative)) {
    return { ok: false, reason: "outside_workspace" };
  }
  return { ok: true, filePath: current, size: stat.size };
}

function limitText(text: string, maxBytes: number, maxLines: number): { text: string; byteCount: number; lineCount: number; truncated: boolean } {
  const lines = text.split("\n");
  const lineLimited = lines.length > maxLines ? lines.slice(0, maxLines).join("\n") : text;
  const byteLimited = limitUtf8Bytes(lineLimited, maxBytes);
  const lineCount = byteLimited.text.length === 0 ? 0 : byteLimited.text.split("\n").length;
  return {
    text: byteLimited.text,
    byteCount: Buffer.byteLength(byteLimited.text, "utf8"),
    lineCount,
    truncated: lines.length > maxLines || byteLimited.truncated,
  };
}

function limitUtf8Bytes(value: string, maxBytes: number): { text: string; truncated: boolean } {
  if (Buffer.byteLength(value, "utf8") <= maxBytes) {
    return { text: value, truncated: false };
  }
  let text = value;
  while (text.length > 0 && Buffer.byteLength(text, "utf8") > maxBytes) {
    text = text.slice(0, -1);
  }
  return { text, truncated: true };
}

function isBinaryBytes(bytes: Uint8Array): boolean {
  for (const byte of bytes) {
    if (byte === 0 || (byte < 9 || (byte > 13 && byte < 32))) {
      return true;
    }
  }
  return false;
}

function hasBinaryLikeText(value: string): boolean {
  return value.includes("\u0000") || /[\u0001-\u0008\u000b\u000c\u000e-\u001f]/.test(value);
}

function isRequiredRequestId(value: unknown): value is string {
  return typeof value === "string" && safeRequestIdPattern.test(value) && !secretTextPattern.test(value);
}

function isSafeId(value: unknown): value is string {
  return typeof value === "string" && safeIdPattern.test(value) && !/assistant|sk-(?:proj-)?/i.test(value);
}

function boundedInteger(value: unknown, min: number, max: number): value is number {
  return typeof value === "number" && Number.isInteger(value) && value >= min && value <= max;
}

function isPlainRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function hasOnlyKeys(record: Record<string, unknown>, allowedKeys: string[]): boolean {
  return Object.keys(record).every((key) => allowedKeys.includes(key));
}

File: apps/plugins/jetbrains/src/main/kotlin/ai/yet/plugin/bridge/ControlledFileRead.kt
package ai.yet.plugin.bridge

import ai.yet.plugin.identity.ProductIdentity
import com.google.gson.JsonObject
import com.google.gson.JsonParser

data class ControlledFileReadRequest(
    val requestId: String,
    val requestIdMintedBy: String,
    val source: String,
    val controlledWorkspaceId: String,
    val runId: String,
    val workspaceRelativePath: String,
    val maxBytes: Int,
    val maxLines: Int,
    val allowBody: Boolean,
)

object ControlledFileRead {
    private const val MaxRequestBytes = 8192
    private val RequestIdRegex = Regex("^(?!.*(?:authorization|bearer|api[_-]?key|token|secret|access[_-]?token|provider[_-]?key|openai[_-]?api[_-]?key|sk-(?:proj-)?[A-Za-z0-9_-]{8,}))[A-Za-z0-9][A-Za-z0-9_.-]{0,127}$", RegexOption.IGNORE_CASE)
    private val SafeIdRegex = Regex("^(?!assistant(?:[._-]|$))(?!.*(?:assistant|sk-(?:proj-)?))[A-Za-z0-9][A-Za-z0-9._-]{0,79}$", RegexOption.IGNORE_CASE)
    private val WorkspacePathRegex = Regex("""^(?!/)(?![A-Za-z]:)(?!~)(?!.*(?:^|/)\.)(?!.*(?:^|/)\.\.(?:/|$))(?!.*//)(?!.*[\:*?"<>|{}\[\]$^+])(?!(?:^|.*/)(?:node_modules|vendor|dist|build|out|target|coverage|__pycache__|generated|tmp|temp|secrets?|credentials?|private)(?:/|$))[A-Za-z0-9][A-Za-z0-9._-]*(?:/[A-Za-z0-9][A-Za-z0-9._-]*)*$""")

    fun isRequestType(raw: String): Boolean = raw.contains("\"gui.controlledAgentFileReadRequest\"")

    fun safeRequestIdFromRaw(raw: String): String? {
        if (raw.toByteArray(Charsets.UTF_8).size > MaxRequestBytes) return null
        val record = parseRecord(raw) ?: return null
        if (record.stringValue("version") != ProductIdentity.bridgeVersion) return null
        if (record.stringValue("type") != "gui.controlledAgentFileReadRequest") return null
        return record.stringValue("requestId")?.takeIf(::isRequestId)
    }

    fun parse(raw: String): ControlledFileReadRequest? {
        if (raw.toByteArray(Charsets.UTF_8).size > MaxRequestBytes) return null
        val record = parseRecord(raw) ?: return null
        if (!record.keySet().all { it in setOf("version", "type", "requestId", "payload") }) return null
        if (record.stringValue("version") != ProductIdentity.bridgeVersion) return null
        if (record.stringValue("type") != "gui.controlledAgentFileReadRequest") return null
        val requestId = record.stringValue("requestId")?.takeIf(::isRequestId) ?: return null
        val payload = record.get("payload")?.takeIf { it.isJsonObject }?.asJsonObject ?: return null
        if (!payload.keySet().all { it in setOf("requestIdMintedBy", "source", "assistantMinted", "controlledWorkspaceId", "runId", "runtimeSessionId", "sessionId", "workspaceRelativePath", "maxBytes", "maxLines", "allowBody", "singleFileOnly", "recursive", "globAllowed", "regexAllowed", "indexingAllowed") }) return null
        val requestIdMintedBy = payload.stringValue("requestIdMintedBy")?.takeIf { it == "gui" || it == "host" } ?: return null
        val source = payload.stringValue("source")?.takeIf { it == "gui" || it == "host" } ?: return null
        if (payload.booleanValue("assistantMinted") != false) return null
        val controlledWorkspaceId = payload.stringValue("controlledWorkspaceId")?.takeIf(::isSafeId) ?: return null
        val runId = payload.stringValue("runId")?.takeIf(::isSafeId) ?: return null
        payload.stringValue("runtimeSessionId")?.takeIf(::isSafeId) ?: if (payload.has("runtimeSessionId")) return null else null
        payload.stringValue("sessionId")?.takeIf(::isSafeId) ?: if (payload.has("sessionId")) return null else null
        val workspaceRelativePath = payload.stringValue("workspaceRelativePath")?.takeIf(::isWorkspaceRelativePath) ?: return null
        val maxBytes = payload.intValue("maxBytes")?.takeIf { it in 1..8192 } ?: return null
        val maxLines = payload.intValue("maxLines")?.takeIf { it in 1..240 } ?: return null
        val allowBody = payload.booleanValue("allowBody") ?: return null
        if (payload.booleanValue("singleFileOnly") != true) return null
        if (payload.booleanValue("recursive") != false) return null
        if (payload.booleanValue("globAllowed") != false) return null
        if (payload.booleanValue("regexAllowed") != false) return null
        if (payload.booleanValue("indexingAllowed") != false) return null
        return ControlledFileReadRequest(
            requestId = requestId,
            requestIdMintedBy = requestIdMintedBy,
            source = source,
            controlledWorkspaceId = controlledWorkspaceId,
            runId = runId,
            workspaceRelativePath = workspaceRelativePath,
            maxBytes = maxBytes,
            maxLines = maxLines,
            allowBody = allowBody,
        )
    }

    fun unsupportedResult(request: ControlledFileReadRequest): String = JsonObject().apply {
        addProperty("version", ProductIdentity.bridgeVersion)
        addProperty("type", "host.controlledAgentFileReadResult")
        addProperty("requestId", request.requestId)
        add("payload", JsonObject().apply {
            addProperty("kind", "controlled_agent_file_read")
            addProperty("version", "2026-06-29")
            addProperty("authority", "bounded_text_file_read")
            addProperty("cloudRequired", false)
            addProperty("executionAllowed", false)
            addProperty("agentStartAllowed", false)
            add("workspace", JsonObject().apply {
                addProperty("controlledWorkspaceId", request.controlledWorkspaceId)
                addProperty("runId", request.runId)
                addProperty("workspaceMode", "worktree")
                addProperty("host", "jetbrains")
                addProperty("privatePathExposed", false)
                addProperty("workspaceLabel", "Controlled worktree")
            })
            add("request", JsonObject().apply {
                addProperty("requestId", request.requestId)
                addProperty("source", request.source)
                addProperty("requestIdMintedBy", request.requestIdMintedBy)
                addProperty("assistantMinted", false)
                addProperty("workspaceRelativePath", request.workspaceRelativePath)
                addProperty("textOnly", true)
                addProperty("maxBytes", request.maxBytes)
                add("budget", JsonObject().apply {
                    addProperty("scope", "single_explicit_file")
                    addProperty("maxBytes", request.maxBytes)
                    addProperty("maxLines", request.maxLines)
                    addProperty("allowBody", false)
                    addProperty("singleFileOnly", true)
                    addProperty("recursive", false)
                    addProperty("globAllowed", false)
                    addProperty("regexAllowed", false)
                    addProperty("indexingAllowed", false)
                    addProperty("budgetLabel", "Metadata only budget")
                })
                addProperty("reason", "JetBrains bounded read is disabled")
            })
            add("policyFlags", JsonObject().apply {
                addProperty("fileReadAllowed", false)
                addProperty("fileWriteAllowed", false)
                addProperty("shellAllowed", false)
                addProperty("gitAllowed", false)
                addProperty("providerAllowed", false)
                addProperty("toolAllowed", false)
                addProperty("hiddenSearchAllowed", false)
                addProperty("indexingAllowed", false)
                addProperty("binaryReadAllowed", false)
                addProperty("symlinkAllowed", false)
                addProperty("autoStartAllowed", false)
                addProperty("autoApplyAllowed", false)
                addProperty("autoRunAllowed", false)
            })
            add("result", JsonObject().apply {
                addProperty("status", "disabled")
                addProperty("cloudRequired", false)
                addProperty("executionAllowed", false)
                addProperty("bodyIncluded", false)
                addProperty("truncated", false)
                addProperty("blockedReason", "read_disabled")
                addProperty("message", "JetBrains bounded read is disabled")
            })
        })
    }.toString()

    fun rejectedResult(requestId: String): String = unsupportedResult(
        ControlledFileReadRequest(
            requestId = requestId,
            requestIdMintedBy = "host",
            source = "host",
            controlledWorkspaceId = "workspace-read-disabled",
            runId = "run-read-disabled",
            workspaceRelativePath = "docs/architecture/read-disabled.md",
            maxBytes = 1,
            maxLines = 1,
            allowBody = false,
        ),
    )

    private fun parseRecord(raw: String): JsonObject? {
        val element = try {
            JsonParser.parseString(raw)
        } catch (_: RuntimeException) {
            return null
        }
        return element.takeIf { it.isJsonObject }?.asJsonObject
    }

    private fun isRequestId(value: String): Boolean = RequestIdRegex.matches(value)

    private fun isSafeId(value: String): Boolean = SafeIdRegex.matches(value)

    private fun isWorkspaceRelativePath(value: String): Boolean = value.length <= 180 && WorkspacePathRegex.matches(value)

    private fun JsonObject.stringValue(name: String): String? {
        val element = get(name) ?: return null
        if (!element.isJsonPrimitive || !element.asJsonPrimitive.isString) return null
        return element.asString
    }

    private fun JsonObject.booleanValue(name: String): Boolean? {
        val element = get(name) ?: return null
        if (!element.isJsonPrimitive || !element.asJsonPrimitive.isBoolean) return null
        return element.asBoolean
    }

    private fun JsonObject.intValue(name: String): Int? {
        val element = get(name) ?: return null
        if (!element.isJsonPrimitive || !element.asJsonPrimitive.isNumber) return null
        return try {
            val value = element.asInt
            if (element.asString == value.toString()) value else null
        } catch (_: RuntimeException) {
            null
        }
    }
}
